sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs

```diff
@@ -18,14 +18,30 @@ internal abstract class OperationInternalBase
         private readonly string _updateStatusScopeName;
         private readonly IReadOnlyDictionary<string, string>? _scopeAttributes;
         private readonly DelayStrategy? _fallbackStrategy;
+        private readonly AsyncLockWithValue<OperationState> _operationStateLock;
 
-        protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response rawResponse, string operationTypeName, IEnumerable<KeyValuePair<string, string>>? scopeAttributes = null, DelayStrategy? fallbackStrategy = null)
+        protected bool HasSucceeded => _operationStateLock.TryGetValue(out OperationState state) && state.HasSucceeded;
+
+        protected RequestFailedException? OperationFailedException => _operationStateLock.TryGetValue(out var state)
+            ? state.OperationFailedException
+            : default;
+
+        protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response rawResponse, string operationTypeName, IEnumerable<KeyValuePair<string, string>>? scopeAttributes = null, DelayStrategy? fallbackStrategy = null, OperationState? finalState = null)
         {
             _diagnostics = clientDiagnostics;
             _updateStatusScopeName = $"{operationTypeName}.UpdateStatus";
-            _scopeAttributes = scopeAttributes?.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
             RawResponse = rawResponse;
-            _fallbackStrategy = fallbackStrategy;
+
+            if (finalState != null)
```

Comment:
To avoid overriding of one final state with another, final state is assigned in constructor instead of separate method.